### PR TITLE
exporter: per node metric exporter

### PR DESCRIPTION
**Pull request by pereman2**
**feature branch pr**
We'll be merging PRs related to the task list below, until then this PR shouldn't be merged. 
----
The number of perf counters keeps growing in each ceph version and mgr's performance is fragile when moving big payloads of data around because of subinterpreters. A way to reduce this load is by moving perf counters away from the mgr. 

**update**: This PR will be the first of many which will be merge onto the feature branch `feature-exporter`. This PR is intended to have all features marked as done in _Next steps_ except TLS support.

**Next steps**:
- [x] List socket files and make use of them
- [x] Create http server to expose GET /metrics for prometheus
- [x] Load perf counters from all daemons and transform to prometheus format
- [x] Containarize
- [x] Introduce a exporter container for cephadm
- [ ] MetricCollector discovery in cephadm in order to configure prometheus to scrape all MetricCollector endpoints (https://github.com/ceph/ceph/pull/45425)
- [ ] Ensure rook support
- [x] Posiblity to expose perf counters so that modules can request through http/whatever
- [ ] **Security**: HTTPS / TLS 🆕 
- [ ] **Security**: HTTP Basic Auth 🆕 
- [x] **Ops**: configuration (via ceph.conf + mon config opts) 🆕 
- [x] **Ops**: application log (errors, warnings, etc.) 🆕 
- [x] **Ops**: server log (HTTP requests) 🆕 
- [ ] **Performance**: HTTP compression 🆕
- [ ] **Performance**: memory/cpu footprint, soak testing (mem-leak) 🆕
- [ ] **Documentation** 🆕
- [ ] **Tests**: Integration tests with teuthology
- [ ] **Tests**: Add unit testing

Proposal for node-exporter:
Screenshot from 2022-03-02 21-30-43

Fixes: https://tracker.ceph.com/issues/55046
Signed-off-by: Pere Diaz Bou <pdiazbou@redhat.com>
Co-authored-by: Avan Thakkar <athakkar@redhat.com>

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/cephadm/cephadm [hunk 1]
```diff
@@ -45,7 +45,7 @@ from pathlib import Path
 FuncT = TypeVar('FuncT', bound=Callable)
 
 # Default container images -----------------------------------------------------
-DEFAULT_IMAGE = 'quay.ceph.io/ceph-ci/ceph:master'
+DEFAULT_IMAGE = 'docker.io/rhcsdashboard/ceph-exporter'
```
Comment (jmolmo) [inline src/cephadm/cephadm:48]: i guess this is just to be used while we are developing the feature.

**Review by jmolmo — COMMENTED** [review]
Great initiative!! Congratulations

Just a couple of doubts about if this is the right solution and about implementation. Maybe  i miss something so i think it is worth to open a discussion about these two points.

**Is this the right solution?**
The exporter task is just to retrieve the data from the different daemons and expose them again as metrics using an http interface.
Probably the hard part of this is to have the daemons data refreshed on-time , and that depends completely of the different Ceph daemons. 

In my view, the right solution is to adopt a efficient and common solution for export metrics in each daemon, and not to create a new ceph metrics exporter.
The ceph metrics exporter you propose has a good thing, which is to provide a single scrape point for node, but as i commented do not solve the main problem we have:
Unified (from the implementation point of view) and efficient metrics exposition in each daemon.
I think that it would be worth to explore the way we are doing that in the different daemons and see how we can optimize the metrics exposition. 

**Implementation:**
Apart of my main concern about if this is the right solution, the current implementation using the beast Beast probably is going to work faster than light, but i think that we do not have the problem in this part, as i commented the hard part is to make the different daemons dump the data fast and frequently.
So probably, if we decide to go ahead with this solution, is better to use a standard approach, i mean, we have several Prometheus client libraries to do the job, why to reinvent the wheel? Here the Beast efficiency probably is not going to give us more advantages than the "normal" http server provided in the different libraries.

**Comment by epuertat** [comment]
> **Is this the right solution?** The exporter task is just to retrieve the data from the different daemons and expose them again as metrics using an http interface. Probably the hard part of this is to have the daemons data refreshed on-time , and that depends completely of the different Ceph daemons.

If I'm not wrong, this approach decouples external scraping from internal metric collection, so it'd be almos instant. If you mean 'metrics freshness', I don't think we need millisecond lag for metrics (the scrape interval usually is 10-15 seconds).

> In my view, the right solution is to adopt a efficient and common solution for export metrics in each daemon, and not to create a new ceph metrics exporter. The ceph metrics exporter you propose has a good thing, which is to provide a single scrape point for node, but as i commented do not solve the main problem we have: Unified (from the implementation point of view) and efficient metrics exposition in each daemon. I think that it would be worth to explore the way we are doing that in the different daemons and see how we can optimize the metrics exposition.

Each (Ceph) daemon is already exposing metrics in a unified/efficient way: perf-counters via daemon socket. This metrics exporter is rather a "metric gateway/broker" that also translates the internal Ceph metrics dialect to the external Prometheus format (other metrics collectors could be supported in the future: InfluxDB, etc).

The reason not to go with a per-daemon exporter was to avoid embedding a webserver inside every Ceph daemon, which could expose the Ceph service to DDoS attacks and other HTTP exploits. Additionally, the cost of managing the service discovery for >1,000s of services is heavier than for ~100 nodes (and nodes are almost static, while services could migrate between nodes).

> **Implementation:** Apart of my main concern about if this is the right solution, the current implementation using the beast Beast probably is going to work faster than light, but i think that we do not have the problem in this part, as i commented the hard part is to make the different daemons dump the data fast and frequently. So probably, if we decide to go ahead with this solution, is better to use a standard approach, i mean, we have several Prometheus client libraries to do the job, why to reinvent the wheel? Here the Beast efficiency probably is not going to give us more advantages than the "normal" http server provided in the different libraries.

I agree with you here (I'm lazy and prefer not to reinvent the wheel every time), but I'm afraid that it's a bit too late to change that.

**Comment by jmolmo** [comment]
> > **Is this the right solution?** The exporter task is just to retrieve the data from the different daemons and expose them again as metrics using an http interface. Probably the hard part of this is to have the daemons data refreshed on-time , and that depends completely of the different Ceph daemons.
> 
> If I'm not wrong, this approach decouples external scraping from internal metric collection, so it'd be almos instant. If you mean 'metrics freshness', I don't think we need millisecond lag for metrics (the scrape interval usually is 10-15 seconds).
> 
Yes, what i mean is that this solution does not act over the real metrics  recollection in each daemon.

> > In my view, the right solution is to adopt a efficient and common solution for export metrics in each daemon, and not to create a new ceph metrics exporter. The ceph metrics exporter you propose has a good thing, which is to provide a single scrape point for node, but as i commented do not solve the main problem we have: Unified (from the implementation point of view) and efficient metrics exposition in each daemon. I think that it would be worth to explore the way we are doing that in the different daemons and see how we can optimize the metrics exposition.
> 
> Each (Ceph) daemon is already exposing metrics in a unified/efficient way: perf-counters via daemon socket. This metrics exporter is rather a "metric gateway/broker" that also translates the internal Ceph metrics dialect to the external Prometheus format (other metrics collectors could be supported in the future: InfluxDB, etc).
> 
> The reason not to go with a per-daemon exporter was to avoid embedding a webserver inside every Ceph daemon, which could expose the Ceph service to DDoS attacks and other HTTP exploits. Additionally, the cost of managing the service discovery for >1,000s of services is heavier than for ~100 nodes (and nodes are almost static, while services could migrate between nodes).
>
Ok. I understand that then this is only a "metric gateway broker". The main function is reduce the number of scrape points, and in the future to act as interface for other metrics scrappers,. 
> > **Implementation:** Apart of my main concern about if this is the right solution, the current implementation using the beast Beast probably is going to work faster than light, but i think that we do not have the problem in this part, as i commented the hard part is to make the different daemons dump the data fast and frequently. So probably, if we decide to go ahead with this solution, is better to use a standard approach, i mean, we have several Prometheus client libraries to do the job, why to reinvent the wheel? Here the Beast efficiency probably is not going to give us more advantages than the "normal" http server provided in the different libraries.
> 
> I agree with you here (I'm lazy and prefer not to reinvent the wheel every time), but I'm afraid that it's a bit too late to change that.

Never is late... :-) ...  the dashboard team decides what is the best approach, for sure. Just to comment that  if this is a metrics broker, to use an abstraction to represent the metrics obtained from the perf counters (or directly the perf counters), and after that to use the different kind of client libraries to produce the final metrics for specific "scrappers" in different endpoints continues being a good idea in my view.

**Comment by MrFreezeex** [comment]
> In my view, the right solution is to adopt a efficient and common solution for export metrics in each daemon, and not to create a new ceph metrics exporter. The ceph metrics exporter you propose has a good thing, which is to provide a single scrape point for node, but as i commented do not solve the main problem we have: Unified (from the implementation point of view) and efficient metrics exposition in each daemon. I think that it would be worth to explore the way we are doing that in the different daemons and see how we can optimize the metrics exposition.

Personally the issue I have with not adopting a standard solution is that we will miss some functionalities available in the Prometheus format. I am thinking of histogram which IIRC are not available with perf counters. There are also exemplars that would provide nice correlation between metrics and traces.

Thinking that if the end goal is to expose a single endpoint per node, it might be nice way to proxy openmetrics/prometheus format all the way to the daemons. Like maybe we could also have a command to ceph daemons or a localhost endpoint that this daemon (or something else that might already exist) fetch and then also expose it... It also might be a way to transition smoothly and collecting both perf counters and the real Prometheus format for some time. If we use a standard lib, there might even be a way to support without too much pain the case where daemon send both metrics and traces with OpenTelemetry at the same time...

**Comment by jmolmo** [comment]
Another thought.. Thinking on how to integrate the ceph daemon exporter in k8s envs. 
Afaik, and if i am not wrong, the "asok" files for the different daemons live in the different pods,and the socket file cannot be found in the host where the pod runs ... so in order to make this  run in k8s envs, all the daemon socket files must be moved to the hosts. 
I think that probably the critical thing here will be updates. Have you check this point with the Rook team?

**Comment by leseb** [comment]
> Another thought.. Thinking on how to integrate the ceph daemon exporter in k8s envs. Afaik, and if i am not wrong, the "asok" files for the different daemons live in the different pods,and the socket file cannot be found in the host where the pod runs ... so in order to make this run in k8s envs, all the daemon socket files must be moved to the hosts. I think that probably the critical thing here will be updates. Have you check this point with the Rook team?
> 
> @leseb , @BlaineEXE, @travisn What do you think?

Makes sense.

**Comment by epuertat** [comment]
> Never is late... :-) ... the dashboard team decides what is the best approach, for sure. Just to comment that if this is a metrics broker, to use an abstraction to represent the metrics obtained from the perf counters (or directly the perf counters), and after that to use the different kind of client libraries to produce the final metrics for specific "scrappers" in different endpoints continues being a good idea in my view.

@jmolmo this wasn't a Dashboard team decision (alone). There were multiple meetings with Ceph devels and users (including a couple of CDM meetings: [[1]](https://tracker.ceph.com/projects/ceph/wiki/CDM_02-FEB-2022) and [[2]](https://tracker.ceph.com/projects/ceph/wiki/CDM_02-MAR-2022)).

> You already heard this from me at CDM a few months ago, but personally the issue I have with not adopting a standard solution is that we will miss some functionalities available in the Prometheus format. I am thinking of histograms/summaries which IIRC are not available with perf counters. There are also exemplars that would provide nice correlation between metrics and traces.

@MrFreezeex In fact Perfcounters do support histograms (`PERFCOUNTER_HISTOGRAM` type), but IMHO a histogram is just a derived data representation, so that should be better left for the visualization tool, where users can adjust the bucketing and the temporal window.

Regarding the embedding of Prometheus exporters inside the daemons, there were [2 cons raised against that proposal](https://tracker.ceph.com/projects/ceph/wiki/CDM_02-FEB-2022):
*  Security & stability: embedding a webserver inside a Ceph daemon might pose a security risk (~75% of 2020 vulnerabilities exploited HTTP/TLS implementation issues) and exposure to DDoS attacks.
* Scalability: the goal of the metrics rework was to deal with the 8,000 OSD scenario (based on the Pawsey testing). That would basically involve +8,000 HTTP endpoints, and while Prometheus advertises it's ability to smoothly scrape up to 10,000 endpoints, beyond 4,000 the system requirements entered the region where sharding/horizontal scaling becomes a must (and I don't remember if that even included HTTPS/compressed endpoints).

> Afaik, and if i am not wrong, the "asok" files for the different daemons live in the different pods,and the socket file cannot be found in the host where the pod runs ... so in order to make this run in k8s envs, all the daemon socket files must be moved to the hosts.
I think that probably the critical thing here will be updates. Have you check this point with the Rook team?

@jmolmo if I'm not wrong @pcuzner @pereman2 and @avanthakkar already discussed this with @travisn and the agreement was to expose the daemon socket files at host level, right?

**Comment by MrFreezeex** [comment]
> @MrFreezeex In fact Perfcounters do support histograms (`PERFCOUNTER_HISTOGRAM` type), but IMHO a histogram is just a derived data representation, so that should be better left for the visualization tool, where users can adjust the bucketing and the temporal window.
> 
> Regarding the embedding of Prometheus exporters inside the daemons, there were [2 cons raised against that proposal](https://tracker.ceph.com/projects/ceph/wiki/CDM_02-FEB-2022):

I know this concerns, even though I don't fully agree, I can understand them... That's why I proposed in my last message an asok command that would dump OpenMetrics format or even a localhost http endpoint (so that this per node exporter may collect it). The perfcounters may support histogram but not the current prometheus module: https://github.com/ceph/ceph/blob/9476907f606922c4b25fc0f063eb0d199ac2ac0c/src/pybind/mgr/prometheus/module.py#L1578 (not sure for this daemon...).

Not sure by what you mean by it "that should be better left for the visualization tool", but histograms are really useful. With them you can compute things like median, certain quantile (i.e.: p95/p99)... And also have a nice way to link traces and metrics, for instance: you have a bucket in your histogram with request between 200 and 500ms, you see that there are some requests in this range with a heatmap in Grafana and then you can click on it to open the trace that correspond to one of these requests.

**Comment by epuertat** [comment]
> Not sure what you mean by "that should be better left for the visualization tool", but histograms are really useful. With them you can compute things like median, certain quantile (i.e.: p95/p99)... And also have a nice way to link traces and metrics, for instance: you have a bucket in your histogram with request between 200 and 500ms, you see that there are some requests in this range with a heatmap in Grafana and then you can click on it to open the trace that correspond to one of these requests. Pretty sure none of these things are possible with the Ceph prometheus monitoring in its current form...

Yeah, I'm not saying that histograms are not useful. I'm just saying that you can still obtain them from the regular time series (counters, long running avgs). For obtaining the 1,2,3,...-sigma quantiles you can still use the `quantile_over_time()` function, and what's a heatmap but a series of quantiles over time? Since v5 Grafana can display histograms/heatmaps from regular time series.

Anyway, if needed, histograms could be exposed by the current mgr/prometheus exporter or this new exporter. Not a big deal.

**Comment by MrFreezeex** [comment]
> Yeah, I'm not saying that histograms are not useful. I'm just saying that you can still obtain them from the regular time series (counters, long running avgs). For obtaining the 1,2,3,...-sigma quantiles you can still use the `quantile_over_time()` function, and what's a heatmap but a series of quantiles over time? Since v5 Grafana can display histograms/heatmaps from regular time series.

I'm not certain that `quantile_over_time` works for the `lat_sum` metrics that Ceph exposes. Like this function does not weight each value. So fo instance, I don't  how you would do that with `ceph_rgw_get_initial_lat_sum` and `ceph_rgw_get_initial_lat_count` to make each entries weighted (also histograms also have some `_sum` and `_count` and `histogram_quantile` doesn't even use those).

For the heatmap visualisation in Grafana indeed it works I didn't know that.

> Anyway, if needed, histograms could be exposed by the current mgr/prometheus exporter or this new exporter. Not a big deal.

Well yes maybe, you may also implement an exemplar equivalent on top of perf counters in the future for tracing too, it doesn't seems that bad. IMO having the native format and using standard libs might still be better to have a better integration in the observability ecosystem in general and be future-proof, but it would probably be good enough either ways...

**Review by jmolmo — APPROVED** [review]

File: src/cephadm/cephadm (hunk 1 above)
Comment (avanthakkar) [reply src/cephadm/cephadm:48]: Yes..the package will be be the part of the default ceph image once merged in master.

**Comment by avanthakkar** [comment]
jenkins test dashboard cephadm

**Comment by avanthakkar** [comment]
jenkins test dashboard cephadm

**Comment by avanthakkar** [comment]
jenkins test dashboard cephadm

File: src/pybind/mgr/cephadm/utils.py [hunk 2]
```diff
@@ -21,7 +21,7 @@ class CephadmNoImage(Enum):
 
 # ceph daemon types that use the ceph container image.
 # NOTE: order important here as these are used for upgrade order
-CEPH_TYPES = ['mgr', 'mon', 'crash', 'osd', 'mds', 'rgw', 'rbd-mirror', 'cephfs-mirror']
+CEPH_TYPES = ['mgr', 'mon', 'crash', 'osd', 'mds', 'rgw', 'rbd-mirror', 'cephfs-mirror', 'ceph-exporter']
```
Comment (sunilangadi2) [inline src/pybind/mgr/cephadm/utils.py:24]: @avanthakkar can we have this `ceph-exporter` in MONITORING_STACK_TYPES as it's intended for Prometheus metrics purpose?
also what version does `ceph-exporter` gets assigned?

File: ceph.spec.in [hunk 3]
```diff
@@ -730,6 +730,15 @@ Requires:	libcephfs2 = %{_epoch_prefix}%{version}-%{release}
 %description -n cephfs-mirror
 Daemon for mirroring CephFS snapshots between Ceph clusters.
 
+%package -n ceph-exporter
```
Comment (ktdreyer) [inline ceph.spec.in:733]:
```suggestion
%package exporter
```

File: src/common/options/ceph-exporter.yaml.in [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+  type: str
+  level: advanced
+  desc: Host ip address where exporter is deployed
+  default: 0.0.0.0
+  services:
+  - ceph-exporter
+- name: exporter_port
+  type: int
+  level: advanced
+  desc: Port to deploy exporter on. Default is 9926
```
Comment (ktdreyer) [inline src/common/options/ceph-exporter.yaml.in:24]: How did you choose port 9926 over any other port?

What will the TLS port be?

(I'd like to know so we could add this to firewalld's definitions.)

There are a lot of definitions in the code for `exporter_port`. Maybe you should rename these to `exporter_http_port` to distinguish from a future `exporter_https_port`

Comment (ktdreyer) [reply src/common/options/ceph-exporter.yaml.in:24]: > How did you choose port 9926 over any other port?

I see now, you registered that `9926` free port at https://github.com/prometheus/prometheus/wiki/Default-port-allocations

Comment (avanthakkar) [reply src/common/options/ceph-exporter.yaml.in:24]: @ktdreyer  TLS isn't supported yet for the exporter server,so for now its just http port I guess.

File: src/pybind/mgr/cephadm/utils.py (hunk 2 above)
Comment (avanthakkar) [reply src/pybind/mgr/cephadm/utils.py:24]: @sunilangadi2 It's already added https://github.com/ceph/ceph/pull/46213/files#diff-d0f7acffbce59b9e36a1479d1b1f32955cd811ab8b0b1aa0018a2003f0c4d73eR5262

Comment (avanthakkar) [reply src/pybind/mgr/cephadm/utils.py:24]: it's a ceph daemon so it will have ceph version

**Comment by avanthakkar** [comment]
**[UPDATE]** : This PR wil ljust contain cephdaemon changes required and not the cephadm changes for introducing new `ceph-exporter` container as we first need a new ceph image based on C++ changes which will contain the exporter package. The cephadm changes are traced in a seperate [PR](https://github.com/ceph/ceph/pull/47383)

**Comment by avanthakkar** [comment]
jenkins retest this please

File: src/common/options/ceph-exporter.yaml.in (hunk 4 above)
Comment (ktdreyer) [reply src/common/options/ceph-exporter.yaml.in:24]: Right, I'm suggesting we rename *now* to avoid confusion

Comment (avanthakkar) [reply src/common/options/ceph-exporter.yaml.in:24]: Done!

**Comment by avanthakkar** [comment]
jenkins test dashboard cephadm

**Comment by avanthakkar** [comment]
jenkins test dashboard cephadm

**Comment by avanthakkar** [comment]
jenkins test windows

**Comment by avanthakkar** [comment]
jenkins test dashboard cephadm

**Comment by avanthakkar** [comment]
jenkins test make check arm64

File: ceph.spec.in (hunk 3 above)
Comment (pereman2) [reply ceph.spec.in:739]: `Exporter` seems a bit too generic given the functionality of this daemon in my opinion.

**Comment by avanthakkar** [comment]
jenkins test make check

**Comment by avanthakkar** [comment]
jenkins test dashboard cephadm

File: ceph.spec.in [hunk 5]
```diff
@@ -736,6 +736,15 @@ Requires:	libcephfs2 = %{_epoch_prefix}%{version}-%{release}
...
+Summary: Daemon for exposing perf counters as Prometheus metrics
+%if 0%{?suse_version}
+Group:		System/Filesystems
+%endif
+Requires:	ceph-base = %{_epoch_prefix}%{version}-%{release}
+%description -n ceph-exporter
+Daemon for exposing perf counters as Prometheus metrics
+
```
Comment (epuertat) [inline ceph.spec.in:747]: I don't see `debian/control` changes, so this won't be available to Debian/Ubuntu. If intentional, then it should be noted somewhere.

**Comment by travisn** [comment]
@pereman2 @jmolmo Would you mind opening an issue in the [Rook repo](https://github.com/rook/rook/issues) to summarize the expected changes needed for Rook? There are a few comments in this PR, we just need to capture them in an issue now. Thanks!
This feature will first be released with Reef?

**Comment by avanthakkar** [comment]
> @pereman2 @jmolmo Would you mind opening an issue in the [Rook repo](https://github.com/rook/rook/issues) to summarize the expected changes needed for Rook? There are a few comments in this PR, we just need to capture them in an issue now. Thanks! This feature will first be released with Reef?

Sure! I can open a issue. The plan is to backport this to quincy as well, but yes I think if we see release branch then it should be Reef I think.

**Pull request merged**